gammapy/modeling/sampler.py

```diff
@@ -0,0 +1,97 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Sampler parameter classes."""
+
+import logging
+import ultranest
+#from .sampler import SamplerLikelihood
+
+__all__ = ["Sampler", "SamplerResult","SamplerLikelihood"]
+
+
+
+
+class Sampler:
+    """Sampler class.
+
+    The sampler class provides a uniform interface to multiple sampler backends.
+    Currently available: "UltraNest", "emcee" in #TODO .
+
+    Parameters
+    ----------
+    backend : {"ultranest"}
+        Global backend used for sampler. Default is "ultranest".
+
+    #TODO : describe all parameters
+    """
+
+    def __init__(self, live_points=100, frac_remain=0.5, log_dir=None, resume=False, step_sampler=False):
+        self._sampler = None
+        self.min_num_live_points = live_points
+        self.frac_remain = frac_remain
+        self.log_dir = log_dir
+        self.step_sampler = step_sampler
+        self.resume = resume
+
+    @staticmethod
+    def _parse_datasets(datasets):
+        from gammapy.datasets import Dataset, Datasets
+
+        if isinstance(datasets, (list, Dataset)):
+            datasets = Datasets(datasets)
+#        return datasets, datasets.parameters.free_parameters
+        return datasets, datasets.parameters
+
+    def run(self, datasets):
+        datasets, parameters = self._parse_datasets(datasets=datasets)
+        parameters = parameters.free_parameters
+
+        # create log likelihood function
+        # need to remove prior penalty
+        like = SamplerLikelihood(function=datasets.stat_sum_no_prior, parameters=parameters)
+        
+        # create ultranest object
+        self._sampler = ultranest.ReactiveNestedSampler(parameters.names, like.fcn, transform=parameters.prior_inverse_cdf,
+                                                        log_dir=self.log_dir, resume=self.resume)
+
+        if self.step_sampler: 
+            print(f'step_sampler={step_sampler} => Using a step sampler' )
+            nsteps = 10
+            #RegionBallSliceSampler
+            self._sampler.stepsampler = ultranest.stepsampler.SliceSampler(
+            nsteps=nsteps,
+            generate_direction=ultranest.stepsampler.generate_mixture_random_direction,
+            adaptive_nsteps=False,
+            # max_nsteps=400
+            )
+        
+        result = self._sampler.run(min_num_live_points=self.min_num_live_points, frac_remain=self.frac_remain)
```

Comment:
could be moved to a function `sample_ultranest` that return `sampler` and `result` similar to what we have for optimize.

Reply:
done